dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/layers/PoolingLayer.java

```diff
@@ -0,0 +1,172 @@
+ * limitations under the License.
+ */
+package edu.snu.dolphin.dnn.layers;
+
+import edu.snu.dolphin.dnn.blas.Matrix;
+import edu.snu.dolphin.dnn.blas.function.Function;
+import edu.snu.dolphin.dnn.conf.LayerConfigurationParameters.*;
+import edu.snu.dolphin.dnn.layerparam.initializer.LayerParameterInitializer;
+import org.apache.reef.tang.annotations.Parameter;
+
+import javax.inject.Inject;
+
+/**
+ * Pooling layer.
+ *
+ * This layer is not learnable.
+ * This layer resizes input matrix spatially, using max pooling or average pooling.
+ * In a forward pass,
+ * max pooling picks maximum value in certain range (kernelHeight * kernelWidth) and these values make up output.
+ * Average pooling gets average of values in certain range (kernelHeight * kernelWidth) and these values make up output.
+ * In a backward pass,
+ * each value of error matrix is the sum of elements in next error matrix
+ * on which had an impact in feedforward step.
+ */
+public final class PoolingLayer extends LayerBase {
+
+  private enum PType {
+    average, max
+  }
+  private final int[] outputShape;
+  private PType poolingType;
+  private int strideHeight;
+  private int strideWidth;
+  private int kernelHeight;
+  private int kernelWidth;
+  private Matrix trackMatrix;
+  private Function poolingFunctions;
+
+  @Inject
+  private PoolingLayer(@Parameter(LayerIndex.class) final int index,
+                       @Parameter(LayerInputShape.class) final String inputShape,
+                       @Parameter(PoolingType.class) final String poolingType,
+                       @Parameter(StrideHeight.class) final int strideHeight,
+                       @Parameter(StrideWidth.class) final int strideWidth,
+                       @Parameter(KernelHeight.class) final int kernelHeight,
+                       @Parameter(KernelWidth.class) final int kernelWidth,
+                       final LayerParameterInitializer layerParameterInitializer) {
+    super(index, inputShape);
+    this.outputShape = setOutputShape();
+    this.strideHeight = strideHeight;
+    this.strideWidth = strideWidth;
+    this.kernelHeight = kernelHeight;
+    this.kernelWidth = kernelWidth;
+
+    try {
+      this.poolingType = PType.valueOf(poolingType);
+    } catch (final IllegalArgumentException illegalArgumentException) {
+      throw new IllegalArgumentException("Illegal pooling type: " + illegalArgumentException);
+    } catch (final NullPointerException nullPointerException) {
+      throw new NullPointerException("Null pointer exception while matching pooling type: " + nullPointerException);
+    }
+  }
+
+  @Override
+  public int[] getOutputShape() {
+    return outputShape;
+  }
+
+  /**
+   * This function computes output shape.
+   * input shape: row * col
+   * output shape: row' * col'
+   * row = (row − kernelHeight) / stride + 1
+   * col = (col − kernelWidth) / stride + 1
+   */
+  private int[] setOutputShape() {
```

Comment:
If I change that function name to `getOutputShape()`, `getOutputShape()` would compute the same value again  whenever that function is called. How about just changing the function name to something like `computeOutputShape()` and keep `getOutputShape()` as now?